File: java/scon-core/src/test/java/io/github/realmlabs/scon/ParserFormatSourceFuzzTest.java
package io.github.realmlabs.scon;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertNotNull;

import com.code_intelligence.jazzer.junit.FuzzTest;
import java.nio.charset.StandardCharsets;
import org.junit.jupiter.api.condition.EnabledIfSystemProperty;

@EnabledIfSystemProperty(named = "scon.fuzz.enabled", matches = "true")
final class ParserFormatSourceFuzzTest {
    @FuzzTest(maxDuration = "30s")
    void formatSource(byte[] data) {
        String source = new String(data, StandardCharsets.UTF_8);
        String formatted;
        try {
            formatted = Scon.formatSource(source);
        } catch (SconException ignored) {
            return;
        }

        assertNotNull(Scon.analyzeSource(formatted).parsed());

        try {
            SconValue original = Scon.parseString(source);
            SconValue roundTrip = Scon.parseString(formatted);
            assertEquals(Scon.formatValue(original), Scon.formatValue(roundTrip));
        } catch (SconException ignored) {
            // If either source fails semantic resolution, formatting parseability
            // remains the invariant under test.
        }
    }
}

File: kotlin/scon-core/src/test/kotlin/io/github/realmlabs/scon/ParserFormatSourceFuzzTest.kt
package io.github.realmlabs.scon

import com.code_intelligence.jazzer.junit.FuzzTest
import org.junit.jupiter.api.condition.EnabledIfSystemProperty
import kotlin.test.assertEquals

@EnabledIfSystemProperty(named = "scon.fuzz.enabled", matches = "true")
@EnabledIfSystemProperty(named = "scon.fuzz.target", matches = "format_source|all")
class ParserFormatSourceFuzzTest {
    @FuzzTest(maxDuration = "30s")
    fun formatSource(data: ByteArray) {
        val source = data.toString(Charsets.UTF_8)
        val formatted = runCatching { formatSource(source) }.getOrNull() ?: return

        parseSource(formatted)

        val original = runCatching { resolveSource(source) }.getOrNull()
        val roundTrip = runCatching { resolveSource(formatted) }.getOrNull()
        if (original != null && roundTrip != null) {
            assertEquals(original, roundTrip)
        }
    }
}
